src/Client/Grpc/GrpcDurableEntityClient.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Google.Protobuf.WellKnownTypes;
+using Microsoft.DurableTask.Client.Entities;
+using Microsoft.DurableTask.Entities;
+using Microsoft.Extensions.Logging;
+using static Microsoft.DurableTask.Protobuf.TaskHubSidecarService;
+using DTCore = DurableTask.Core;
+using P = Microsoft.DurableTask.Protobuf;
+
+namespace Microsoft.DurableTask.Client.Grpc;
+
+/// <summary>
+/// The client for entities.
+/// </summary>
+class GrpcDurableEntityClient : DurableEntityClient
+{
+    readonly TaskHubSidecarServiceClient sidecarClient;
+    readonly GrpcDurableTaskClient durableTaskClient;
+    readonly ILogger logger;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="GrpcDurableEntityClient"/> class.
+    /// </summary>
+    /// <param name="durableTaskClient">The durable Task client.</param>
+    /// <param name="sidecarClient">The client for the GRPC connection to the sidecar.</param>
+    /// <param name="logger">The logger for logging client requests.</param>
+    public GrpcDurableEntityClient(GrpcDurableTaskClient durableTaskClient, TaskHubSidecarServiceClient sidecarClient, ILogger logger)
+        : base(durableTaskClient.Name)
+    {
+        this.durableTaskClient = durableTaskClient;
+        this.sidecarClient = sidecarClient;
+        this.logger = logger;
+    }
+
+    /// <inheritdoc/>
+    public override async Task SignalEntityAsync(EntityInstanceId id, string operationName, object? input = null, SignalEntityOptions? options = null, CancellationToken cancellation = default)
+    {
+        Guid requestId = Guid.NewGuid();
+        DateTimeOffset? scheduledTime = options?.SignalTime;
+
+        P.SignalEntityRequest request = new P.SignalEntityRequest()
+        {
+            InstanceId = id.ToString(),
+            RequestId = Google.Protobuf.ByteString.FromStream(new MemoryStream(requestId.ToByteArray())),
+            Name = operationName,
+            Input = this.durableTaskClient.DataConverter.Serialize(input),
+            ScheduledTime = scheduledTime.HasValue ? Timestamp.FromDateTimeOffset(scheduledTime.Value.ToUniversalTime()) : default,
+        };
+
+        // TODO this.logger.LogSomething
+        await this.sidecarClient.SignalEntityAsync(request, cancellationToken: cancellation);
+    }
+
+    /// <inheritdoc/>
+    public override async Task<EntityMetadata?> GetEntityAsync(EntityInstanceId id, bool includeState = false, CancellationToken cancellation = default)
+    {
+        P.GetEntityRequest request = new P.GetEntityRequest()
+        {
+            InstanceId = id.ToString(),
+            IncludeState = includeState,
+        };
+
+        P.GetEntityResponse response = await this.sidecarClient.GetEntityAsync(request, cancellationToken: cancellation);
+
+        return response == null ? null : this.ToEntityMetadata(response.Entity, includeState);
```

Comment:
I don't think this will ever be null, it will most likely throw before that. Or the server has a bad implementation, in which case we should just fail.

Reply:
Good catch. This should instead be testing `response.Exists`.